code/modules/projectiles/guns/flamer/flamer.dm

```diff
@@ -226,6 +229,58 @@
 
 	new /obj/flamer_fire(to_fire, create_cause_data(initial(name), user), R, max_range, current_mag.reagents, flameshape, target, CALLBACK(src, PROC_REF(show_percentage), user), fuel_pressure, fire_type)
 
+/obj/item/weapon/gun/flamer/proc/unleash_foam(atom/target, mob/living/user)
+	last_fired = world.time
+	if(!current_mag || !current_mag.reagents || !current_mag.reagents.reagent_list.len)
+		return
+
+	var/source_turf = get_turf(user)
+	var/foam_range = 6 // the max range the foam will travel
+	var/distance = 0 // the distance traveled
+	var/use_multiplier = 3 // if you want to increase the ammount of foam drained from the tank
+	var/datum/reagent/chemical = current_mag.reagents.reagent_list[1]
+
+	var/turf/turfs[] = get_line(user, target, FALSE)
+	var/turf/first_turf = turfs[1]
+	var/ammount_to_use = (min(turfs.len, foam_range) * use_multiplier) // the ammount of units that will be drained from the tank
+	for(var/turf/turf in turfs)
+
+		if(current_mag.reagents.total_volume < ammount_to_use)
+			foam_range = round(current_mag.reagents.total_volume / use_multiplier)
+
+		if(distance >= foam_range)
+			break
+
+		if(turf.density)
+			break
+		else
+			var/obj/effect/particle_effect/foam/checker = new()
+			var/atom/blocked = LinkBlocked(checker, source_turf, turf)
+			if(blocked)
+				break
+
+		if(turf == first_turf) // this is so the first foam tile doesn't expand and touch the user
+			var/datum/effect_system/foam_spread/foam = new()
+			foam.set_up(0.5, turf, metal_foam = FOAM_METAL_TYPE_IRON)
+			foam.start()
+		else
+			var/datum/effect_system/foam_spread/foam = new()
+			foam.set_up(1, turf, metal_foam = FOAM_METAL_TYPE_IRON)
+			foam.start()
+		sleep(2)
+
+		distance++
+
+	// this is likely not needed but I want to be on the safe side
+	current_mag.reagents.total_volume = clamp(current_mag.reagents.total_volume -= (distance * use_multiplier), 0, current_mag.reagents.maximum_volume)
+
+	chemical.volume = clamp(chemical.volume -= (distance * use_multiplier), 0, current_mag.reagents.maximum_volume)
```

Comment:
Refactor these. It is needlessly complex what you are intending to occur here by having assignments within proc calls compounded into another assignment. Retest after reworking this because it will likely behave differently.

In otherwords break these operations into several different lines rather than trying to single line all the changes.

Reply:
done, hopefully this is better

Reply:
Still unresolved. Don't do assignment inside of code like this:
![image](https://github.com/cmss13-devs/cmss13/assets/76988376/9b758969-ef20-45cb-a00a-53feafb6acee)

`thing -= some_value`
is
`thing = thing - some_value`

So what you have is a single line of this logic:
```js
var/temp = chemical.volume = chemical.volume - ammount_used
chemical.volume = max(temp, 0)
```

Reply:
oops forgot to close this